### PR TITLE
move alternatives to manual mode even if paths match

### PR DESCRIPTION
**Pull request by triddle-maxmind**
##### SUMMARY
Currently the alternatives module does not perform in an intuitive way when invoked to set an alternative that matches the current configuration but auto mode is in use. If the path specified for an alternative matches the path currently in use but the alternative is in auto mode ansible will make no change to the system. I believe it makes more sense to move alternatives to manual mode in this condition.

My rationale is as follows:

  1. Given a task that installs a package that participates in alternatives
  2. Given a task that follows the first sets the path in alternatives for the package
  3. Given a task that installs a package which is higher priority than the previous alternative
  4. Then the alternatives configuration will not match what was specified in number 2

I believe that when using Ansible to configure an alternative it is a surprise to have the alternative change even when installing a package that provides an alternative of higher priority. As an example of this consider what happens when the hypothetical play is executed twice:

  1. On first run the first package is installed and used by alternatives
  2. On first run the second package is installed and becomes the current alternative
  3. On the second run no packages are installed but the desired alternative will be configured and alternatives will move into manual mode.

This does not appear to be desired behavior.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
Alternatives module

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

### CODE REVIEW

**Comment by bcoca** [comment]
Since this breaks current behaviour, I would favor for it to be controlled from a toggle instead 'force=no|yes' which would handle this 'auto' case.

File: lib/ansible/modules/system/alternatives.py [hunk 1]
```diff
@@ -116,14 +116,16 @@ def main():
...
     if rc == 0:
         # Alternatives already exist for this link group
-        # Parse the output to determine the current path of the symlink and
-        # available alternatives
+        # Parse the output to determine the current path of the symlink,
+        # available alternatives and the mode
         current_path_regex = re.compile(r'^\s*link currently points to (.*)$',
                                         re.MULTILINE)
         alternative_regex = re.compile(r'^(\/.*)\s-\spriority', re.MULTILINE)
+        mode_regex = re.compile(r'^' + re.escape(name) + ' - (.*) mode')
```
Comment (DavidWittman) [inline lib/ansible/modules/system/alternatives.py]: This regex doesn't appear to match on RHEL 7.

``` bash
$ update-alternatives --display editor
editor - status is auto.
 link currently points to /usr/bin/vim
/usr/bin/vim - priority 10
/usr/bin/nano - priority 0
Current `best' version is /usr/bin/vim.
```

**Review by DavidWittman — CHANGES_REQUESTED** [review]

**Comment by DavidWittman** [comment]
> Since this breaks current behaviour, I would favor for it to be controlled from a toggle instead 'force=no|yes' which would handle this 'auto' case.

I agree this breaks current behavior, but I also think the current behavior is broken. When an alternative is set, the user should expect it to remain set. Manual mode is the only way to ensure that. Is a `force` flag  necessary even when the change fixes what could be classified as a bug?

**Comment by triddle-maxmind** [comment]
>  Is a force flag necessary even when the change fixes what could be classified as a bug?

If my two cents are worth anything I believe using a force option to change this behavior makes the user perform steps to solve a bug.

> This regex doesn't appear to match on RHEL 7.

I can work on a replacement regex that'll match under a RHEL and Debian based OS.

**Comment by triddle-maxmind** [comment]
I updated the PR so a function is used to identify when auto mode is active for a given alternative. The function uses a cascade of regular expressions and raises an exception if none of them match.

**Comment by aklemp** [comment]
There is the argument that the fix would break current behavior. But the module is marked being in state `preview` and explicitly saying that the interface is not guaranteed to be backwards compatible. However, in addition to this fix I propose further states `selected` and `auto` to make the desired state explicit and only install the alternative for state `present` without changing the default configuration. See also #24351.

**Comment by triddle-maxmind** [comment]
I'm willing to work on #24351 in tandem with this bug fix but I'm unsure if it should be broken up into two different PRs or not. Would it be better to include the changes from #24351 in this pull request or submit a PR for #24351 that is based on the branch this bugfix comes from? Perhaps there is some other way to do it?

**Comment by aklemp** [comment]
A single PR would reduce the effort for the person merging the PR. But I'm not sure about the development effort. Does this bug fix require some changes that will be undone and changed by #24351? What is the overhead of creating an additional PR and keeping it aligned with this bugfix if they are on the same branch?

What about PR #23444 that is introducing the state parameter but only with `absent`/`present`? It seems to be ready for being merged and someone signalled working on the additional states as well.

Either way, thank you for working on it. :-)

**Pull request closed without merging**